include/drivers/regulator.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2019-2020 Peter Bigot Consulting, LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief API for regulators controlled by a GPIO.
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_GPIO_REGULATOR_H_
+#define ZEPHYR_INCLUDE_DRIVERS_GPIO_REGULATOR_H_
+
+#include <zephyr/types.h>
+#include <drivers/gpio.h>
+#include <sys/onoff.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Driver-specific API functions to support regulator control.
+ */
+struct regulator_driver_api {
+	int (*enable)(struct device *dev, struct onoff_client *cli);
+	int (*disable)(struct device *dev);
+};
+
+/**
+ * @brief Enable a regulator.
+ *
+ * Reference-counted request that a regulator be turned on.  This is
+ * an asynchronous operation; if successfully initiated the result
+ * will b e communicated through the @p cli parameter.
+ *
+ * @note This function is *isr-ok* and *pre-kernel-*ok.
+ *
+ * @param reg a regulator device
+ *
+ * @param cli used to notify the caller when the attempt to turn on
+ * the regulator has completed.
+ *
+ * @return non-negative on successful initiation of the request.
+ * Negative values indicate failures from onoff_request() or
+ * individual regulator drivers.
+ */
+static inline int regulator_enable(struct device *reg,
+				   struct onoff_client *cli)
+{
+	const struct regulator_driver_api *api =
+		(const struct regulator_driver_api *)reg->driver_api;
+
+	return api->enable(reg, cli);
+}
+
+/**
+ * @brief Disable a regulator.
+ *
+ * Turn off a regulator after a previous regulator_enable() completed
+ * successfully.
+ *
+ * This must be invoked exactly once for each successful enable call.
+ *
+ * @note This function is *isr-ok*.
+ *
+ * @param reg a regulator device
+ *
+ * @return non-negative on successful completion of the release
```

Comment:
No.  This was a design trade-off in the context of developing the on-off service.  If it's necessary to detect when the transition completes that can be done in other ways.

I suspect it will be necessary, so this may have to change.

Reply:
In general you can only know that you have started to turn it off. Of course you can look at datasheets and some of them will have a 'ready' bit which tells you that the power is stable (or fully off).

IMO it is probably enough to know that we have completed 'asking' for it to turn off. A client could add a udelay if needed.

For turning on, I think we need to know that the power is fully stable, since until it is we cannot advance.